osgbuild/fetch_sources.py

```diff
@@ -330,14 +364,14 @@ def fancy_source_error(meta_type, explicit_type, handler, args, kw, e):
 
     xtype = "type" if explicit_type else "implicit type"
     log.error("Error processing source line of %s '%s'" % (xtype, meta_type))
-    varnames = handler.__code__.co_varnames
-    fn_argcount = handler.__code__.co_argcount
-    minargs = fn_argcount - len(handler.__defaults__)
-    reqargs = varnames[:minargs]
-    maxargs = varnames.index('ops')
-    posargs = varnames[:maxargs]
-    posargs_provided = posargs[:len(args)]
-    dupe_args = set(posargs_provided) & set(kw)
+    varnames = handler.__code__.co_varnames     # arg names for handler func
```

Comment:
Is `__code__` just the function header, or does it contain the code block as well? Why is it called `varnames` if it's the names of the _args_?

Reply:
So, for reference: https://docs.python.org/3/reference/datamodel.html

Under Data Model:
- [3.2. The standard type hierarchy](https://docs.python.org/3/reference/datamodel.html#the-standard-type-hierarchy)
  - Callable types
    - User-defined functions
      - Special attributes:
        | Attribute | Meaning |
        | --- | --- |
        |`__code__` | The code object representing the compiled function body.|
  - Internal types
    - Code objects
      - Special read-only attributes:
        - `co_argcount` is the total number of positional arguments (including positional-only arguments and arguments with default values);
        - `co_varnames` is a tuple containing the names of the local variables (starting with the argument names);
        - `co_code` is a string representing the sequence of bytecode instructions;
        ...

So yes, `func.__code__` contains the bytecode (`co_code`) as well as various function metadata attributes.

> Why is it called `varnames` if it's the names of the _args_?

>```     varnames = handler.__code__.co_varnames```

I thought it was obvious that `varnames` here is being used as shorthand for `...co_varnames`.

(Technically, the reason why python calls the one `co_varnames` and the other `co_argcount` is that varnames includes (first) the variable names that represent the arguments to the function, followed by the local variable names as they appear in the function.  Though in this case we are only using it for the names of the function's positional parameters, as determined by `co_argcount`.)

Reply:
Carl, if I have to look at the Python language spec to figure out why something works, I'm not gonna say "oh duh" :)

Reply:
I said "with any luck"  :-p

But what i meant was, the variable names offer themselves as documentation.

I am not going to include all of the relevant blocks from the python language data model documentation in my comments, but the hope was that the variable names explain what is going on in a straightforward manner.  So much so that adding a comment to explain it (with any luck!) might sound redundant.

So if you see a line that says
```
reqargs = varnames[:minargs]
```
the hope is that, if somehow the meaning of this statement is not obvious, then further reading
```
# names of required arguments
```
will make it so clear that you not only understand the concept, but also realize that the code itself tried to explain the same thing to you, and in your evrika moment you exclaim, "oh duh."

Reply:
I've tweaked the comment for `varnames` which hopefully clarifies things.